File: CompanionApps/iOS/Sources/Server/CompanionServer.swift
import Foundation
import GRPCCore
import GRPCNIOTransportHTTP2

/// Manages the gRPC server lifecycle for the iOS companion.
///
/// Uses grpc-swift v2 with NIO HTTP/2 transport. The server runs inside an
/// XCUITest target — the test method starts the server and blocks, keeping the
/// companion alive to accept commands from the host.
final class CompanionServer: Sendable {
    let port: Int
    private let provider: CompanionServiceProvider

    init(bridge: XCUITestBridge, port: Int = 22087) {
        self.port = port

        let touch = TouchHandler(bridge: bridge)
        let gesture = GestureHandler(bridge: bridge)
        let text = TextHandler(bridge: bridge)
        let accessibility = AccessibilityHandler(bridge: bridge)

        provider = CompanionServiceProvider(
            touch: touch,
            gesture: gesture,
            text: text,
            accessibility: accessibility
        )
    }

    /// Start the gRPC server and block until shutdown.
    ///
    /// Binds to the loopback interface — the host driver reaches the simulator
    /// via the shared host network, and on a real device the host port-forwards
    /// over USB. Exposing the companion on every interface would let any device
    /// on the network drive XCUITest gestures (taps, types, screenshots).
    func run() async throws {
        let transport = HTTP2ServerTransport.Posix(
            address: .ipv4(host: "127.0.0.1", port: port),
            transportSecurity: .plaintext
        )

        let server = GRPCServer(
            transport: transport,
            services: [provider]
        )

        print("[CompanionServer] Listening on port \(port)")
        try await server.serve()
    }
}

File: CompanionApps/Android/app/src/androidTest/java/com/amoo/companion/server/CompanionServer.kt
package com.amoo.companion.server

import com.amoo.companion.bridge.UIAutomatorBridge
import com.amoo.companion.handlers.AccessibilityHandler
import com.amoo.companion.handlers.GestureHandler
import com.amoo.companion.handlers.TextHandler
import com.amoo.companion.handlers.TouchHandler
import io.grpc.Server
import io.grpc.netty.NettyServerBuilder
import java.net.InetAddress
import java.net.InetSocketAddress

/**
 * Manages the gRPC server lifecycle for the Android companion.
 *
 * The server runs inside an Android instrumentation test. The test method
 * starts the server and blocks, keeping the companion alive to accept
 * commands from the host.
 */
class CompanionServer(
    private val bridge: UIAutomatorBridge,
    private val port: Int = 22088
) {
    private val touch = TouchHandler(bridge)
    private val gesture = GestureHandler(bridge)
    private val text = TextHandler(bridge)
    private val accessibility = AccessibilityHandler(bridge)

    private var server: Server? = null

    /**
     * Start the gRPC server on the configured port.
     *
     * Implementation note: The actual gRPC server setup will use grpc-java
     * (io.grpc:grpc-netty) with the proto-generated stubs compiled into
     * this target.
     */
    fun start() {
        // Bind the ipv4 wildcard explicitly rather than `forPort(port)`. On this image both land
        // on a dual-stack `tcp6 [::]:<port>` socket (that is just how the JDK renders an ipv4
        // wildcard while `java.net.preferIPv4Stack` is false), and that socket already accepts the
        // ipv4-mapped connections `adb forward` makes — its device end always dials ipv4
        // `127.0.0.1:<port>`. The explicit `0.0.0.0` documents that requirement and keeps the
        // listener reachable even if a future image flips `preferIPv4Stack` or sets
        // `IPV6_V6ONLY`, either of which would otherwise leave a `forPort` listener refusing every
        // forwarded connection while still showing a LISTEN socket in `netstat` — from the host,
        // indistinguishable from a companion that is slow to start.
        server = NettyServerBuilder.forAddress(InetSocketAddress(InetAddress.getByName("0.0.0.0"), port))
            .addService(CompanionServiceImpl(touch, gesture, text, accessibility))
            .intercept(DescriptionBackfillInterceptor())
            .build()
            .start()
        println("[CompanionServer] Ready on port $port")
    }

    fun stop() {
        server?.shutdownNow()
        println("[CompanionServer] Shutting down")
    }

    fun blockUntilShutdown() {
        server?.awaitTermination()
    }
}
